renku/core/commands/providers/dataverse.py

```diff
@@ -270,7 +278,180 @@ def remote_url(self):
+            author_template = Template(AUTHOR_METADATA_TEMPLATE)
+            author = author_template.substitute(
+                name=creator.name, affiliation=creator.affiliation
+            )
+            authors.append(json.loads(author))
+
+            contact_template = Template(CONTACT_METADATA_TEMPLATE)
+            contact = contact_template.substitute(
+                name=creator.name, email=creator.email
+            )
+            contacts.append(json.loads(contact))
+
+        return authors, contacts
+
+
+@attr.s
+class _DataverseDeposition:
+    """Dataverse record for deposit."""
+
+    server_url = attr.ib(default=None, kw_only=True)
+    access_token = attr.ib(kw_only=True)
+    dataset_pid = attr.ib(default=None)
+
+    DATASET_CREATE_PATH = 'dataverses/{dataverseName}/datasets'
+    FILE_UPLOAD_PATH = 'datasets/:persistentId/add'
+    DATASET_PUBLISH_PATH = 'datasets/:persistentId/actions/:publish'
+
+    def create_dataset(self, dataverse_name, metadata):
+        """Create a dataset in a given dataverse."""
+        api_path = self.DATASET_CREATE_PATH.format(
+            dataverseName=dataverse_name
+        )
+        url = self._make_url(api_path=api_path)
+
+        response = self._post(url=url, json=metadata)
+        self._check_response(response)
+
+        self.dataset_pid = response.json()['data']['persistentId']
+
+        return response
+
+    def upload_file(self, full_path, path_in_dataset):
+        """Upload a file to a previously-created dataset."""
+        if self.dataset_pid is None:
+            raise errors.ExportError('Dataset not created.')
+
+        url = self._make_url(
+            self.FILE_UPLOAD_PATH, persistentId=self.dataset_pid
+        )
+
+        params = {'directoryLabel': str(path_in_dataset.parent)}
+        data = dict(jsonData=json.dumps(params))
+
+        files = {'file': (path_in_dataset.name, open(full_path, 'rb'))}
+
+        response = self._post(url=url, data=data, files=files)
+        self._check_response(response)
+
+        return response
+
+    def publish_dataset(self):
+        """Publish a previously-created dataset."""
+        if self.dataset_pid is None:
+            raise errors.ExportError('Dataset not created.')
+
+        url = self._make_url(
+            self.DATASET_PUBLISH_PATH,
+            persistentId=self.dataset_pid,
+            type='major'
+        )
+
+        response = self._post(url=url)
+        self._check_response(response)
+
+        return response
+
+    def _make_url(self, api_path, **query_params):
+        """Create URL for creating a dataset."""
+        url_parts = urlparse.urlparse(self.server_url)
+        path = pathlib.posixpath.join(DATAVERSE_API_PATH, api_path)
+
+        query_params = urllib.parse.urlencode(query_params)
+        url_parts = url_parts._replace(path=path, query=query_params)
+        return urllib.parse.urlunparse(url_parts)
+
+    def _post(self, url, json=None, data=None, files=None):
```

Comment:
We should make this more robust, with #902 merged there is a new utility for retrying.
With it you could:
```
with retry() as session:
    session.post(...)
```